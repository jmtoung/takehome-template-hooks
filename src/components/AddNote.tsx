import React, { useState } from 'react';
import { Button } from 'react-bootstrap';
import { API } from '../api';
import { useInput } from '../hooks/useInput';
import { NoteForm } from './NoteForm';

export const AddNote: React.FunctionComponent<{
  total: number;
  page: string;
  fetchNotes: () => {};
}> = (props) => {
  const [isCreating, setIsCreating] = useState<boolean>(false);
  const { value: title, onChange: changeTitle, reset: resetTitle } = useInput(
    ''
  );
  const { value: body, onChange: changeBody, reset: resetBody } = useInput('');

  const handleSubmit = async (event: React.FormEvent<HTMLFormElement>) => {
    event.preventDefault();
    await API.addNote(title, body);
    resetTitle();
    resetBody();
    setIsCreating(false);

    const { page, total } = props;
    if (parseInt(page, 10) === Math.ceil(total / 10)) {
      props.fetchNotes();
    }
  };

  if (!isCreating) {
    return (
      <Button className="create-note" onClick={() => setIsCreating(true)}>
        Create Note
      </Button>
    );
  }
  return (
    <NoteForm
      text="Create Note"
      title={title}
      body={body}
      handleSubmit={(e) => handleSubmit(e)}
      handleBodyChange={changeBody}
      handleTitleChange={changeTitle}
    />
  );
};
