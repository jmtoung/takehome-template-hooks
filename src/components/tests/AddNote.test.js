import React from 'react';
import AddNote from '../AddNote';
import toJSON from 'enzyme-to-json';
import { shallow } from 'enzyme';

describe('AddNote', () => {
  it('renders addNote button view correctly.', () => {
    const wrapper = shallow(<AddNote />);
    expect(toJSON(wrapper)).toMatchSnapshot();
  });

  it('calls updateAddNoteView and displays form view.', () => {
    const wrapper = shallow(<AddNote />);
    const changeView = jest.spyOn(wrapper.instance(), 'updateAddNoteView');

    wrapper.find('.create-note').simulate('click');

    expect(changeView).toHaveBeenCalledTimes(1);
    expect(toJSON(wrapper)).toMatchSnapshot();
  });

  it('updates state when handleChange is called.', () =>  {
    const wrapper = shallow(<AddNote />);

    const handleChange = jest.spyOn(wrapper.instance(), 'handleChange');

    const form = document.createElement('form');
    const input = document.createElement('input');
    form.appendChild(input);
    input.onchange = handleChange.bind(wrapper);
    input.value = 'This is the title';
    input.name = 'title';

    const event = new Event('change');
    input.dispatchEvent(event);

    expect(handleChange).toHaveBeenCalledTimes(1);

    expect(wrapper.state()).toEqual({
      title: 'This is the title',
      body: '',
      isCreating: false
    });

  });

  it('sends new note to api when form is submitted.', (done) => {
    const wrapper = shallow(<AddNote />);
    wrapper.find('.create-note').simulate('click');
    wrapper.setState({title: 'title', body: 'body'});

    const mockSuccessResponse = {id: '1', title: 'This is the created title', body: 'This is the created body.'};
    const mockJsonPromise = Promise.resolve(mockSuccessResponse);
    const mockFetchPromise = Promise.resolve({
      json: () => mockJsonPromise,
    });
    jest.spyOn(global, 'fetch').mockImplementation(() => mockFetchPromise);


    var submitEvent = new Event('submit');
    wrapper.instance().handleSubmit(submitEvent);

    expect(global.fetch).toHaveBeenCalledTimes(1);
    expect(global.fetch).toHaveBeenCalledWith("https://5ec581652a4ba000163d2084.mockapi.io/api/v1/notes", {"body": "{\"title\":\"title\",\"body\":\"body\"}", "headers": {"Content-Type": "application/json"}, "method": "POST"});

    process.nextTick(() => { // after note is created, state is reset.
      expect(wrapper.state()).toEqual({
        title: '',
        body: '',
        isCreating: false
      });
    });

    global.fetch.mockClear();
    done();
  });

});
