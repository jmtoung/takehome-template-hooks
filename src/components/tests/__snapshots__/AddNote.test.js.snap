// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddNote calls updateAddNoteView and displays form view. 1`] = `
<NoteForm
  body=""
  handleChange={[Function]}
  handleSubmit={[Function]}
  text="Add Note"
  title=""
/>
`;

exports[`AddNote renders addNote button view correctly. 1`] = `
<Button
  active={false}
  className="create-note"
  disabled={false}
  onClick={[Function]}
  type="button"
  variant="primary"
>
  Create Note
</Button>
`;
